dm: update dm worker intro (#21465)
pingcap/docs: dm/dm-worker-intro.md

# DM-worker Introduction

DM-worker is a component of TiDB Data Migration (DM) that executes tasks to dump and replicate data from MySQL/MariaDB to TiDB.

## Key Concepts

* A DM-worker can perform a full export of data from the MySQL source and then transition to reading the MySQL binlog for continuous, incremental replication.

> * The DM-worker is the execution engine for the tasks and subtasks received from the DM-master. It dumps data from one MySQL source instance, acts as a replication client reading the binlog events, performs data transformation and filtering, stores data in a local relay log, applies data to the downstream target TiDB, and reports the status back to the DM-master.

Author: ```suggestion
- The DM-worker is the execution engine for tasks and subtasks received from the DM-master. It dumps data from one MySQL source instance, acts as a replication client reading the binlog events, performs data transformation and filtering, stores data in a local relay log, applies data to the downstream target TiDB, and reports the status back to the DM-master.
```

# DM-worker Introduction

DM-worker is a component of TiDB Data Migration (DM) that executes tasks to dump and replicate data from MySQL/MariaDB to TiDB.

## Key Concepts

> * A DM-worker can perform a full export of data from the MySQL source and then transition to reading the MySQL binlog for continuous, incremental replication.

Author: ```suggestion
- A DM-worker can perform a full export of data from the MySQL source and then transition to reading the MySQL binlog for continuous, incremental replication.
```

> - The DM-worker is the execution engine for tasks and subtasks received from the DM-master. It dumps data from one MySQL source instance, acts as a replication client reading the binlog events, performs data transformation and filtering, stores data in a local relay log, applies data to the downstream target TiDB, and reports the status back to the DM-master.

Reviewer: I think relay log is optional and default to false, so this implication of every DM-worker persisting binlog events locally could be misleading.

Author: Updated the overview to describe the default path without relay logging. The relay-log section now states that relay logging is optional and disabled by default, and explains when to enable it and its resource and latency tradeoffs.

- If a worker instance goes offline, DM-master can automatically reschedule its tasks to another available worker to resume the data replication. Note that this does not apply during a full export/import phase.

> - A single DM-worker process connects to **one** upstream source database at a time. To migrate from multiple sources, such as when merging sharded tables, you must run multiple DM-worker processes.

Reviewer 2: ```suggestion
- A single DM-worker process connects to **one** upstream source database instance at a time. To migrate from multiple sources, such as when merging sharded tables, you must run multiple DM-worker processes.
```

> * If a worker instance goes offline, DM-master can automatically reschedule its tasks to another available worker to resume the data replication. Note that this does not apply during a full export/import phase.
> * A single DM-worker process connects to **one** upstream source database at a time. To migrate from multiple sources, such as when merging sharded tables, you must run multiple DM-worker processes.

Author: ```suggestion
- If a worker instance goes offline, DM-master can automatically reschedule its tasks to another available worker to resume the data replication. Note that this does not apply during a full export/import phase.
- A single DM-worker process connects to **one** upstream source database at a time. To migrate from multiple sources, such as when merging sharded tables, you must run multiple DM-worker processes.
```

# DM-worker Introduction

> DM-worker is a component of TiDB Data Migration (DM) that executes tasks and subtasks assigned by DM-master. For a full and incremental migration, it dumps data from one MySQL-compatible source instance, then reads the source binlog as a replication client, transforms and filters events, and applies them to the target TiDB cluster. DM-master queries DM-worker for source and subtask status.

Author: Updated the overview to include the load step explicitly. The sentence remains scoped to a full and incremental migration, and now describes dump, load, then binlog replication in sequence.

> DM-worker is a component of TiDB Data Migration (DM) that executes tasks and subtasks assigned by DM-master. For full and incremental migration, it dumps data from one MySQL-compatible source instance and loads the dumped data into the target TiDB cluster. It then reads the source binlog as a replication client, transforms and filters events, and applies them to the target. DM-master queries DM-worker for source and subtask status.

Reviewer 2: ```suggestion
DM-worker is a component of TiDB Data Migration (DM) that executes tasks and subtasks assigned by DM-master. For full and incremental migration, it dumps data from one MySQL-compatible source instance and loads the dumped data into the target TiDB cluster. It then reads the source binlog as a replication client, transforms and filters events, and applies them to the target. DM-master queries DM-worker for the status of sources and subtasks.
```

> DM-worker is a component of TiDB Data Migration (DM) that executes tasks and subtasks assigned by DM-master. For a full and incremental migration, it dumps data from one MySQL-compatible source instance and loads the dumped data into the target TiDB cluster. It then reads the source binlog as a replication client, transforms and filters events, and applies them to the target. DM-master queries DM-worker for source and subtask status.

Reviewer 2: ```suggestion
DM-worker is a component of TiDB Data Migration (DM) that executes tasks and subtasks assigned by DM-master. For full and incremental migration, it dumps data from one MySQL-compatible source instance and loads the dumped data into the target TiDB cluster. It then reads the source binlog as a replication client, transforms and filters events, and applies them to the target. DM-master queries DM-worker for source and subtask status.
```

> DM-worker is a component of TiDB Data Migration (DM) that executes tasks and subtasks assigned by DM-master. For full and incremental migration, it dumps data from one MySQL-compatible source instance and loads the dumped data into the target TiDB cluster. It then reads the source binlog as a replication client, transforms and filters events, and applies them to the target. DM-master queries DM-worker for the status of sources and subtasks.

Reviewer 2: ```suggestion
DM-worker is a component of TiDB Data Migration (DM) that executes tasks and subtasks assigned by DM-master. For a full and incremental migration, it dumps data from one MySQL-compatible source instance and loads the dumped data into the target TiDB cluster. It then reads the source binlog as a replication client, transforms and filters events, and applies them to the target. DM-master queries DM-worker for the status of sources and subtasks.
```

## Key concepts

- If a worker instance goes offline, DM-master can automatically reschedule its tasks to another available worker to resume the data replication. Note that this does not apply during a full export/import phase.
- A single DM-worker process connects to **one** upstream source database instance at a time. To migrate from multiple sources, such as when merging sharded tables, you must run multiple DM-worker processes.

> **Note:**
>

> > A DM-worker is a MySQL binlog client, not a standby database replica server. It reads and replays data from a MySQL source to a TiDB target. To replicate data from a source TiDB cluster, use [TiCDC](/ticdc/ticdc-overview.md).

Author: Updated both phrases to `MySQL-compatible` so the note consistently includes MariaDB and other supported MySQL-compatible sources.

## DM-worker processing units

Depending on its task mode, a DM-worker subtask runs the dump, load, and binlog replication processing units. DM-worker can also run an optional relay log processing unit for its bound source.

### Relay log

> Relay logging is optional and disabled by default. When enabled, DM-worker stores upstream binlog events on local disk before the binlog replication processing unit reads them. Enable relay logging when a long full migration or blocked task might outlast upstream binlog retention, or when multiple tasks for the same source need to reuse one binlog stream. Relay logging consumes disk, I/O, and CPU resources and can increase replication latency. For configuration and operational details, see [DM relay log](/dm/relay-log.md).

Reviewer 2: ```suggestion
Relay logging is optional and disabled by default. When enabled, DM-worker stores upstream binlog events on the local disk before the binlog replication processing unit reads them. Enable relay logging if a long-running full migration or a blocked task might outlast upstream binlog retention period, or if multiple tasks for the same source need to share a single binlog stream. Relay logging consumes disk, I/O, and CPU resources, and can increase replication latency. For configuration and operational details, see [DM relay log](/dm/relay-log.md).
```

> Relay logging is optional and disabled by default. When enabled, DM-worker stores upstream binlog events on the local disk before the binlog replication processing unit reads them. Enable relay logging if a long-running full migration or a blocked task might outlast upstream binlog retention period, or if multiple tasks for the same source need to share a single binlog stream. Relay logging consumes disk, I/O, and CPU resources, and can increase replication latency. For configuration and operational details, see [DM relay log](/dm/relay-log.md).

Reviewer 2: ```suggestion
Relay log is optional and disabled by default. When enabled, DM-worker stores upstream binlog events on the local disk before the binlog replication processing unit reads them. Enable relay log if a long-running full migration or a blocked task might outlast upstream binlog retention period, or if multiple tasks for the same source need to share a single binlog stream. Relay logging consumes disk, I/O, and CPU resources, and can increase replication latency. For configuration and operational details, see [DM relay log](/dm/relay-log.md).
```

| `REPLICATION SLAVE` | Global |
| `REPLICATION CLIENT` | Global |

To grant these privileges, execute the following statement:

```sql
GRANT RELOAD, REPLICATION SLAVE, REPLICATION CLIENT ON *.* TO 'your_user'@'your_wildcard_of_host';
GRANT SELECT ON `db1`.* TO 'your_user'@'your_wildcard_of_host';
```

#### MariaDB version 10.5 or higher

Starting from [MariaDB 10.5](https://mariadb.com/docs/release-notes/community-server/old-releases/mariadb-10-5-series/what-is-mariadb-105), the `REPLICATION CLIENT` privilege was renamed and split into more granular privileges. The user must have the following privileges:

| Privilege | Scope | Description |
|:---|:---|
| `SELECT` | Tables | Required for full data export. |
| `RELOAD` | Global | Required for `FLUSH TABLES WITH READ LOCK`. |
| `BINLOG MONITOR` | Global | Renamed from `REPLICATION CLIENT`; allows monitoring the binlog. |
| `REPLICATION SLAVE` | Global | Allows reading binlog events. |

> | `REPLICATION SLAVE ADMIN` | Global | Allows managing replication status (e.g., `SHOW SLAVE STATUS`). |

Author: ```suggestion
| `REPLICATION SLAVE ADMIN` | Global | Allows managing replication status (for example, `SHOW SLAVE STATUS`). |
```

> | `REPLICATION MASTER ADMIN`| Global | Allows monitoring the master (e.g., `SHOW SLAVE HOSTS`). |

Author: ```suggestion
| `REPLICATION MASTER ADMIN`| Global | Allows monitoring the master (for example, `SHOW SLAVE HOSTS`). |
```

To grant these privileges, execute the following statement:

```sql

> CREATE USER 'tidb-dm'@'%' IDENTIFIED BY 'MyPassw0rd!';
> GRANT SELECT, RELOAD, BINLOG MONITOR, REPLICATION SLAVE, REPLICATION SLAVE ADMIN, REPLICATION MASTER ADMIN ON *.* TO 'tidb-dm'@'%';
> FLUSH PRIVILEGES;

Author: ```suggestion
GRANT RELOAD, BINLOG MONITOR, REPLICATION SLAVE, REPLICATION SLAVE ADMIN, REPLICATION MASTER ADMIN ON *.* TO 'your_user'@'your_wildcard_of_host';
GRANT SELECT ON db1.* TO 'your_user'@'your_wildcard_of_host';
```

```sql
GRANT RELOAD, BINLOG MONITOR, REPLICATION SLAVE, REPLICATION SLAVE ADMIN, REPLICATION MASTER ADMIN, REPLICA MONITOR ON *.* TO 'your_user'@'your_wildcard_of_host';
GRANT SELECT ON `db1`.* TO 'your_user'@'your_wildcard_of_host';
```

> **Note:**
>
> Because MariaDB reports these privileges differently from MySQL, `dmctl check-task` might report privilege errors even when the account has the required privileges.
>
> For DM v8.5.6, if the precheck returns `[code=26005] fail to check synchronization configuration` for the replication privilege, dump privilege, or dump connection number check, add only the following items to the task configuration file:
>
> ```yaml
> ignore-checking-items:
>   - replication_privilege
>   - dump_privilege
>   - conn_number
> ```
>
> This workaround skips only the three checks affected by the MariaDB privilege parser. Before using it, manually verify the corresponding privileges and connection limit. For more information, see [DM precheck](/dm/dm-precheck.md).

> **Note:**
>

> > If the dump unit fails with `Error 1227` while querying the MariaDB `INNODB_TABLESPACES_SCRUBBING` or `INNODB_TABLESPACES_ENCRYPTION` table, grant `SUPER` to let the dump unit read the required metadata. `SUPER` is a broad privilege, so add it only when this exact error occurs and your security policy permits it.

Author: Added `PROCESS` as the normal MariaDB privilege for InnoDB metadata queries and to the applicable `GRANT` examples. `SUPER` remains a conditional fallback only for the exact reproduced `Error 1227` path. In the DM v8.5.6 smoke test, MariaDB 10.4.34, 10.5.1, and 10.5.2 still required `SUPER` even with `PROCESS`; the tested 10.5.9, 10.6.13, and 10.11.16 releases completed without it.

| `REPLICATION SLAVE` | Global |
| `REPLICATION CLIENT` | Global |

To grant these privileges, execute the following statement:

```sql
GRANT RELOAD, REPLICATION SLAVE, REPLICATION CLIENT ON *.* TO 'your_user'@'your_wildcard_of_host';
GRANT SELECT ON `db1`.* TO 'your_user'@'your_wildcard_of_host';
```

For a full data export from MariaDB earlier than 10.5.2, also grant `PROCESS` so that the dump unit can query InnoDB metadata:

```sql
GRANT PROCESS ON *.* TO 'your_user'@'your_wildcard_of_host';
```

#### MariaDB 10.5.2 to 10.5.8

> Starting from [MariaDB 10.5.2](https://mariadb.com/docs/release-notes/community-server/old-releases/10.5/10.5.2), the `REPLICATION CLIENT` privilege is renamed and split into more granular privileges. For MariaDB 10.5.2 to 10.5.8, the user must have the following privileges:

Author: Corrected. The text now states that `REPLICATION CLIENT` was renamed to `BINLOG MONITOR`, while `SUPER` was split into granular privileges used by replication statements. The agreed 10.5.2 and 10.5.9 version boundaries remain unchanged.

```sql
GRANT PROCESS, RELOAD, BINLOG MONITOR, REPLICATION SLAVE, REPLICATION SLAVE ADMIN, REPLICATION MASTER ADMIN, REPLICA MONITOR ON *.* TO 'your_user'@'your_wildcard_of_host';
GRANT SELECT ON `db1`.* TO 'your_user'@'your_wildcard_of_host';
```

> **Note:**
>
> Because MariaDB reports these privileges differently from MySQL, `dmctl check-task` might report privilege errors even when the account has the required privileges.
>
> For DM v8.5.6, if the precheck returns `[code=26005] fail to check synchronization configuration` for the replication privilege, dump privilege, or dump connection number check, add only the following items to the task configuration file:
>
> ```yaml
> ignore-checking-items:
>   - replication_privilege
>   - dump_privilege
>   - conn_number
> ```
>
> This workaround skips only the three checks affected by the MariaDB privilege parser. Before using it, manually verify the corresponding privileges and connection limit. For more information, see [DM precheck](/dm/dm-precheck.md).

> **Note:**
>

> > On some older MariaDB releases, `PROCESS` is not sufficient for the dump unit to query InnoDB metadata. With DM v8.5.6, this behavior occurs when the dump unit queries `INNODB_TABLESPACES_SCRUBBING` on MariaDB 10.4.34, or `INNODB_TABLESPACES_ENCRYPTION` on MariaDB 10.5.1 and 10.5.2. In the same smoke tests, MariaDB 10.5.9, 10.6.13, and 10.11.16 complete without `SUPER`. If the dump unit returns `Error 1227 (42000): Access denied; you need (at least one of) the SUPER privilege(s) for this operation`, grant `SUPER`. Because `SUPER` is a broad privilege, grant it only when this exact error occurs and your security policy permits it.

Reviewer 2: ```suggestion
> On some older MariaDB releases, `PROCESS` is not sufficient for the dump unit to query InnoDB metadata. With DM v8.5.6, this behavior occurs when the dump unit queries `INNODB_TABLESPACES_SCRUBBING` on MariaDB 10.4.34, or `INNODB_TABLESPACES_ENCRYPTION` on MariaDB 10.5.1 and 10.5.2. In the same smoke tests, MariaDB 10.5.9, 10.6.13, and 10.11.16 complete without `SUPER`.
>
> If the dump unit returns the following error, grant `SUPER`. Because `SUPER` is a broad privilege, grant it only when this exact error occurs and your security policy permits it.
>
> ```
> Error 1227 (42000): Access denied; you need (at least one of) the SUPER privilege(s) for this operation
> ```
```

| `REPLICATION SLAVE` | Global |
| `REPLICATION CLIENT` | Global |

To grant these privileges, execute the following statement:

```sql
GRANT RELOAD, REPLICATION SLAVE, REPLICATION CLIENT ON *.* TO 'your_user'@'your_wildcard_of_host';
GRANT SELECT ON `db1`.* TO 'your_user'@'your_wildcard_of_host';
```

#### MariaDB version 10.5 or higher

> Starting from [MariaDB 10.5](https://mariadb.com/docs/release-notes/community-server/old-releases/mariadb-10-5-series/what-is-mariadb-105), the `REPLICATION CLIENT` privilege was renamed and split into more granular privileges. The user must have the following privileges:

Reviewer: If I understand this correctly, these split privileges were introduced in 10.5.2, and starting from 10.5.9, `SHOW SLAVE/REPLICA STATUS` requires `REPLICA MONITOR`, which is missing from the current `GRANT`. Could you be more specific about the applicable version ranges?

Author: Updated the guidance to use the exact MariaDB ranges: earlier than 10.5.2, 10.5.2 to 10.5.8, and 10.5.9 or later. The 10.5.9 or later grant now includes `REPLICA MONITOR`, including the `SLAVE MONITOR` display name from `SHOW GRANTS`. I also replaced the blanket precheck bypass with the three scoped items used by DM v8.5.6.
